rust/src/treefile.rs

```diff
@@ -576,10 +576,42 @@ impl TreefileExternals {
 /// array elements.
 fn whitespace_split_packages(pkgs: &[String]) -> Vec<String> {
     pkgs.iter()
-        .flat_map(|pkg| pkg.split_whitespace().map(String::from))
+        .flat_map(|element| split_whitespace_unless_quote(element).map(String::from))
         .collect()
 }
 
+// Helper for whitespace_split_packages().
+// Splits a String by whitespace unless substring is wrapped between single quotes
+// and returns split Strings in an Iterator.
+fn split_whitespace_unless_quote(element: &String) -> std::vec::IntoIter<String> {
+    let mut quoted_pkgs: Vec<String> = vec![];
+    let mut to_whitespace_split: Vec<String> = vec![];
+    let mut start_index = 0;
+    let mut looping_over_quoted_pkg = false;
+    for (i, c) in element.chars().enumerate() {
```

Comment:
I'm OK if we try to handle the fully general case but...I would also be totally fine if we just required that users only include a single package on a line if using version qualifiers.

IOW, 
```
packages:
  - kernel grub2
  - "'selinux-policy > 1.5'"
  - "'podman > 2.0'"
```

but not:
```
packages:
  - kernel grub2
  - "'selinux-policy > 1.5' 'podman > 2.0'"
```

Reply:
If if we make that a requirement, we can simply check:

```
if element.starts_with('\'') {
   ...
}
```
which would simplify all of this code a lot.

Reply:
It would indeed. I consulted with Jonathan about this and he recommended solving for the general case. I originally had a hacky solution using Rust's library that would shorten the code by a lot, but it felt a bit unintuitive and difficult to understand what was going on.

Reply:
> I consulted with Jonathan about this and he recommended solving for the general case

There might've a misunderstanding. It's 100% fine with me if you want to go the easy route (I initially suggested it myself in https://github.com/coreos/rpm-ostree/issues/2138#issuecomment-646667591). That said, if you feel like tackling the generic case, that's totally cool with me too! :)

Reply:
Ah, I see. I'm okay with either way, as well :)
I'll leave it as is for now, but let me know if I should change it!

Reply:
If we're keeping the code as is I'd need to think hard about whether it's correct; it looks *plausible* - the unit tests would help a lot too particularly if we have failure conditions like missing quotes, etc.

If we only handle the simple case I'd be much more confident it's correct.